include/nbl/builtin/hlsl/bxdf/reflection/iridescent.hlsl

```diff
@@ -0,0 +1,211 @@
+        anisocache_type anisocache;
+        sample_type s = ggx_aniso.generate(anisotropic_interaction_type::create(interaction), u, anisocache);
+        cache = anisocache.iso_cache;
+        return s;
+    }
+
+    scalar_type pdf(NBL_CONST_REF_ARG(query_type) query, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    {
+        struct SGGXDG1Query
+        {
+            using scalar_type = scalar_type;
+
+            scalar_type getNdf() NBL_CONST_MEMBER_FUNC { return ndf; }
+            scalar_type getG1over2NdotV() NBL_CONST_MEMBER_FUNC { return G1_over_2NdotV; }
+
+            scalar_type ndf;
+            scalar_type G1_over_2NdotV;
+        };
+
+        SGGXDG1Query dg1_query;
+        ndf_type ggx_ndf = __base.getNDF();
+        dg1_query.ndf = __base.__D(cache);
+
+        const scalar_type devsh_v = query.getDevshV();
+        dg1_query.G1_over_2NdotV = ggx_ndf.G1_wo_numerator_devsh_part(interaction.getNdotV(_clamp), devsh_v);
+
+        measure_transform_type dualMeasure = __base.template __DG1<SGGXDG1Query>(dg1_query);
+        return dualMeasure.getMicrofacetMeasure();
+    }
+
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(query_type) query, NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    {
+        scalar_type _pdf = pdf(query, interaction, cache);
+
+        spectral_type quo = hlsl::promote<spectral_type>(0.0);
+        if (_sample.getNdotL() > numeric_limits<scalar_type>::min && interaction.getNdotV() > numeric_limits<scalar_type>::min)
+        {
+            struct SGGXG2XQuery
+            {
+                using scalar_type = scalar_type;
+
+                scalar_type getDevshV() NBL_CONST_MEMBER_FUNC { return devsh_v; }
+                scalar_type getDevshL() NBL_CONST_MEMBER_FUNC { return devsh_l; }
+                BxDFClampMode getClampMode() NBL_CONST_MEMBER_FUNC { return _clamp; }
+
+                scalar_type devsh_v;
+                scalar_type devsh_l;
+                BxDFClampMode _clamp;
+            };
+
+            ndf_type ggx_ndf = __base.getNDF();
+            
+            SGGXG2XQuery g2_query;
+            g2_query.devsh_v = query.getDevshV();
+            g2_query.devsh_l = query.getDevshL();
+            g2_query._clamp = _clamp;
+            const scalar_type G2_over_G1 = ggx_ndf.template G2_over_G1<SGGXG2XQuery, sample_type, isotropic_interaction_type, isocache_type>(g2_query, _sample, interaction, cache);
+        
+            fresnel_type f = __base.getFresnel();
+            f.absCosTheta = cache.getLdotH();
+            const spectral_type reflectance = f();
+            quo = reflectance * G2_over_G1;
+        }
+
+        return quotient_pdf_type::create(quo, _pdf);
+    }
+
+    SCookTorrance<Config, ndf_type, fresnel_type, measure_transform_type> __base;
+};
```

Comment:
you could have just added a Delta NDF, and then used a CookTorrance with Iridescent Fresnel